docs/userguide/core-workflow.md

```diff
@@ -97,73 +108,180 @@ struct State {
+modify this state itself, but it may change between render passes. This public state is called
+`Props`.
+
+In Swift, the props are simply defined as properties of the struct implementing Workflow itself. In
+Kotlin, the `Workflow` interface defines a separate `PropsT` type parameter. (This additional type
+parameter is necessary due to Kotlin’s lack of the `Self` type that Swift workflow’s
+`workflowDidChange` method relies upon.)
+
+```Swift tab=
+TK
+```
+
+```Kotlin tab=
+data class Input(
+  val playerXName: String
+  val playerOName: String
+)
+```
+
+## Workflows are advanced by `WorkflowAction`s
+
+Any time something happens that should advance a workflow – a UI event, a network response, a
+child's output event – actions are used to perform the update. For example, a workflow may respond
+to UI events by mapping those events into a type conforming to/implementing `WorkflowAction`. These
+types implement the logic to advance a workflow by:
+
+- Advancing to a new state
+- (Optionally) emitting an output event up the tree.
+
+`WorkflowAction`s are typically defined as enums with associated types (Swift) or sealed classes
+(Kotlin), and can include data from the event – for example, the ID of the item in the list that was
+clicked.
+
+Side effects such as logging button clicks to an analytics framework are also typically performed in
+actions.
+
+If you're familiar with React/Redux, `WorkflowAction`s are essentially reducers.
+
+## Workflows can emit output events up the hierarchy to their parent
+
+When a workflow is advanced by an action, an optional output event can be sent up the workflow
+hierarchy. This is the opportunity for a workflow to notify its parent that something has happened
+(and the parent's opportunity to respond to that event by dispatching its own action, continuing up
+the tree as long as output events are emitted).
 
-Immediately after starting up, or after a state transition occurs, a workflow will have its
-`render(state:context:)` method called. This method is responsible for creating and returning a
-value of type `Rendering`. You can think of `Rendering` as the "external state" of the workflow.
-While a workflow's internal state may contain more detailed or comprehensive state, the `Rendering`
-(external state) is a type that is useful outside of the workflow.
+## Workflows produce an external representation of their state via `Rendering`
+
+Immediately after starting up, or after a state transition occurs, a workflow will have its `render`
+method called. This method is responsible for creating and returning a value of type `Rendering`.
+You can think of `Rendering` as the "external published state" of the workflow, and the `render`
+function as a map of `Input` + `State` + childrens' `Rendering`s -> `Rendering`. While a workflow's
+internal state may contain more detailed or comprehensive state, the `Rendering` (external state) is
+a type that is useful outside of the workflow.
 
```

Comment:
Consider putting something like:

> Because a workflow’s `render` method may be called by infrastructure for a variety of reasons, it’s important to not perform side effects when rendering. Event-based side effects should use Actions and state-based side effects should use Workers.

Reply:
> …important to not perform side effects when rendering — render methods _must_ be idempotent.